### PR TITLE
Feat/create providers user pool

### LINKED ISSUE #201: An existing licensee can log in
- separate cognito user pool is stood up for provider users
- configure with app client for UI to authenticate (mostly copy from staff login)
- frontend needs client ID for user pool and login URL 
- user pool configuration must restrict which attributes a user can modify (can modify given name, family name, email)
- deliverable: provider successfully logged in

### PR DESCRIPTION
**Pull request by landonshumway-ia**
We need to have a separate user pool to track licensee users (AKA providers) so they can authenticate with the app
separately from staff users.

### Requirements List
- No new requirements for this change.

### Description List
- Update UserPool common construct to allow child classes to define list of standard attributes.
- Added ProviderUsers construct with email configuration, standard attributes, and custom `providerId` and `compact` attributes.

### Testing List
- `yarn test:unit:all` should run without errors or warnings
- `yarn serve` should run without errors or warnings
- `yarn build` should run without errors or warnings
- Code review

Closes #201

### CODE REVIEW

**Comment by jlkravitz** [comment]
- [ ] review the pull request to get oriented
	- [ ] read the description of the pull request, which should summarize the changes made
	- [ ] read through every task on the Scrum board that's encompassed by this pull request
	- [ ] read the description of the commits that comprise the pull request
- [ ] skim all new code, in the context of existing code, looking for problems (knowing that the vast majority of new code will be covered by tests)
- [ ] review all tests
	- [ ] methodically review all new tests for correctness, quality of naming
	- [ ] look at code coverage of tests
	- [ ] determine what code isn’t tested, review that rigorously
- [ ] review documentation to ensure that it matches changes
- [ ] provide comments on the pull request on GitHub, as necessary
	- [ ] for comments that are specific to a particular line of code, comment on those specific lines
	- [ ] for comments that are more general, attach the comment to a random line in `README.md` (as opposed to commenting on the pull request itself), to be able to use GitHub's ability to thread discussions on those comments
- [ ] run a security audit of dependencies (e.g. `npm audit` and `pip audit`) to ensure that there are no vulnerabilities that will be deployed to production (as opposed to vulnerabilities that only have an impact on the development environment)

File: backend/compact-connect/common_constructs/user_pool.py [hunk 1]
```diff
@@ -107,11 +117,8 @@ def add_ui_client(self, callback_urls: List[str], ui_scopes: List[OAuthScope] =
             enable_token_revocation=True,
             generate_secret=False,
             refresh_token_validity=Duration.days(30),
-            # If you provide no attributes at all here, it will default
-            # to making _all_ attributes writeable, so if we want to limit writes,
-            # we have to provide at least _one_ that the client _can_ write.
-            write_attributes=ClientAttributes().with_standard_attributes(email=True),
```
Comment (jlkravitz) [inline backend/compact-connect/common_constructs/user_pool.py:113]: Nit: Can we ensure we don't run into this issue in the future, e.g. by asserting that `write_attributes` (in the new code) is non-empty?

File: backend/compact-connect/common_constructs/user_pool.py
Comment (jlkravitz) [inline backend/compact-connect/common_constructs/user_pool.py:1]: General Comment: Can we update `_inspect_persistent_stack` in `test_app.py` to check that the Provider user pool is configured with the expected callbacks (as we do with the Staff user pool). I'm not sure if there are other existing user pool tests, but if there are, it'd be great to update those, too.

**Review by jlkravitz — COMMENTED** [review]
Two small nits but looks great!

File: backend/compact-connect/common_constructs/user_pool.py (hunk 1 above)
Comment (landonshumway-ia) [reply backend/compact-connect/common_constructs/user_pool.py:113]: Now that we are adding type hints for these parameters, the developer should not be able to enter a None type for either the read or write attributes, so this issue should not occur again. The only time that CDK will default to including all attributes is if you pass in None for the argument.

File: backend/compact-connect/common_constructs/user_pool.py
Comment (landonshumway-ia) [reply backend/compact-connect/common_constructs/user_pool.py:1]: Great callout. I've added several new test cases to the CDK stack to verify that the provider user pool is configured with expected properties.

File: backend/compact-connect/common_constructs/user_pool.py (hunk 1 above)
Comment (landonshumway-ia) [reply backend/compact-connect/common_constructs/user_pool.py:113]: @jusdino corrected me on this section, it appears that even if the developer enters `write_attributes=ClientAttributes().with_standard_attributes()` all the attributes would be set to writable, which doesn't align with what the documentation says for that class 😞. I have added some cdk tests that explicitly check the read and write attributes that are set for the two user pool app clients, which would catch if someone attempts to change it to an empty list. Other than that, I'm not sure how you would make an assertion on the object itself to see what parameters it was called with.

Comment (jlkravitz) [reply backend/compact-connect/common_constructs/user_pool.py:113]: Those tests sound great- thanks for doing that! And ditto on the sad face ... what an API :-)

File: backend/compact-connect/common_constructs/user_pool.py
Comment (jlkravitz) [reply backend/compact-connect/common_constructs/user_pool.py:1]: Amazing, thank you!

**Review by jlkravitz — COMMENTED** [review]
little nit but otherwise looking good! thanks for the additional tests and comments

**Comment by landonshumway-ia** [comment]
Hi @jlkravitz, I'm not seeing the nit you mentioned anywhere in the review. Not sure if something is wrong with what I'm seeing in the UI or if the nit was not posted. Just want to confirm what change is needed. Thanks.

**Comment by jlkravitz** [comment]
Weird...let me see...

**Review by jlkravitz — APPROVED** [review]
I think I had a comment that I then decided to delete :) This looks great.

**Pull request merged**